# Threding Tutorial 

    1. Structure of threading code
    2. Sequential program execution of code
    3. parallal execution of code using threading 
    4. comparision of both approches
    5. threading code using custom threading class
    6. threading code execution explanation
    7. threading when a function retruns a value in the code.
        1. differnt ways to handel retrun values.
    8. threading use cases.
        1. IO bound operations
        2. paralla execution of tasks
        3. handlling concurrent use request
        4. handling background tasks

### Structure followed in the examples

In [1]:
import time

def fn():    # function to called in main
    pass

def main():   # main function
    pass

if __name__ == "__main__":   # run only if run as main file not when it gets imported
    main()

### structure of threading

In [3]:
import threading

def fn(n):
    print(n*n)
 
    
def main(N):
    
    threads = []
    for i in range(N):
        num = (i+1)*100000
        t = threading.Thread(target= fn, args=(num,))   # create thread
        t.start()                   # start thread
        threads.append(t)           # append to threds list
        
    for t in threads:
        t.join()                   # join each thread after completion of executions
        
if __name__ == "__main__":
    main(3)
    

### Example of sequential execution of program

In [21]:
import time

def fn(n):
    sum_of_sqr = 0
    for i in range(n):
        sum_of_sqr += i**2
    

def sleepy(sec):
    time.sleep(sec)


def main(N, secs):

    start = time.time()
    for i in range(N):
        num = (1+i)*10000000
        fn(num)
    tm2 = round(time.time() - start,1)
    print('time taken by sum of square fn: ', tm2, 'seconds')

    start = time.time()
    for i in range(1, secs):
        sleepy(i)    
    tm1 = round(time.time() - start,1)
    print('time taken by sleep time fn: ', tm1, 'seconds')

    print(f"total execution time of both fn: {tm1+tm2} seconds")

    

if __name__ == "__main__":   # run only if run as main file not when it gets imported
    main(4, 6)

time taken by sum of square fn:  8.0 seconds
time taken by sleep time fn:  15.0 seconds
total execution time of both fn: 23.0 seconds


### Example of parallal execution of the program

In [18]:
import time
import threading



def fn(n):
    sum_of_sqr = 0
    for i in range(n):
        sum_of_sqr += i**2
    #print(sum_of_sqr)

def sleepy(sec):
    time.sleep(sec)


def main(N, secs):

    start = time.time()

    current_threads = []
    for i in range(N):
        num = (1+i)*10000000
        t = threading.Thread(target= fn, args=(num,))
        t.start()
        current_threads.append(t)
    for i in range(len(current_threads)):
        current_threads[i].join()

    tm1 = round(time.time() - start,1)
    print("Time taken by sum of square fn: ", tm1,'seconds')

    start = time.time()

    current_threads = []
    for second in range(1, secs):
        t = threading.Thread(target= sleepy, args=(second,))
        t.start()
        current_threads.append(t)
    for i in range(len(current_threads)):
        current_threads[i].join()

    tm2 = round(time.time() - start,1)
    print('time taken by sleep time fn: ', tm2, 'seconds')


    print(f"total execution time of both fn: {tm1+tm2} seconds")

    

if __name__ == "__main__":   # run only if run as main file not when it gets imported
    main(4, 6)

Time taken by sum of square:  8.1 seconds
time taken by sleep time:  5.0 seconds
total execution time: 13.1 seconds


 ##### Sequential Execution:
    
    time taken by sum of square fn:  8.0 seconds   (cpu compute intensive task)
    time taken by sleep time fn:  15.0 seconds      (IO bound task)
    total execution time of both fn: 23.0 seconds
    --------------------------------------------------------------------
##### Parllal Execution:
    
    Time taken by sum of square:  8.1 seconds  (cpu compute intensive task)
    time taken by sleep time:  5.0 seconds      (IO bound task)
    total execution time: 13.1 seconds
    -----------------------------------------------------------------
    
##### Conclusion:
    
    the threading optimise the execution time for IO bound task effectively as compare to cpu compute intensive task.
    
    

### same code above using custom threading class

In [25]:
import threading

class FnClass(threading.Thread):
    def __init__(self, n):
        threading.Thread.__init__(self)
        self.n = n

    def run(self):
        sm = 0
        for i in range(self.n):
            sm += i**2


# Custom thread class for sleeping
class SleepyThread(threading.Thread):
    def __init__(self, sec):
        threading.Thread.__init__(self)
        self.sec = sec

    def run(self):
        time.sleep(self.sec)

def main(N,secs):
    start = time.time()
    sum_threads = []
    
    for i in range(N):
        num = (1+i)* 10000000
        t = FnClass(num)
        t.start()
        sum_threads.append(t)

    for t in sum_threads:
        t.join()

    tm1 = round(time.time()- start,1)
    print("Time taken by sum of square fn: ", tm1, 'seconds')

    # Timing the sleep function
    start = time.time()
    sleep_threads = []

    for second in range(1, secs):
        t = SleepyThread(second)
        t.start()
        sleep_threads.append(t)

    for t in sleep_threads:
        t.join()

    tm2 = round(time.time() - start, 1)
    print('Time taken by sleep time fn: ', tm2, 'seconds')

    print(f"Total execution time of both functions: {tm1 + tm2} seconds")



if __name__ == "__main__":
    main(4, 6)

Time taken by sum of square fn:  8.0 seconds
Time taken by sleep time fn:  5.0 seconds
Total execution time of both functions: 13.0 seconds


### Threading Concept Explanation

In [22]:
import time
import threading

def fn(n, thread_id):
    print(f"Thread {thread_id} (fn) started.")
    sum_of_sqr = 0
    for i in range(n):
        sum_of_sqr += i**2
    print(f"Thread {thread_id} (fn) finished with sum: {sum_of_sqr}")

def sleepy(sec, thread_id):
    print(f"Thread {thread_id} (sleepy) started, sleeping for {sec} seconds.")
    time.sleep(sec)
    print(f"Thread {thread_id} (sleepy) finished.")

def main(N, secs):

    start = time.time()

    # Sum of squares calculation with threading
    current_threads = []
    for i in range(N):
        num = (1 + i) * 10000000
        thread_id = f"fn-{i + 1}"
        t = threading.Thread(target=fn, args=(num, thread_id))
        t.start()
        current_threads.append(t)
    
    for t in current_threads:
        t.join()

    tm1 = round(time.time() - start, 1)
    print("Time taken by sum of square fn: ", tm1, 'seconds\n')

    start = time.time()

    # Sleep function with threading
    current_threads = []
    for second in range(1, secs):
        thread_id = f"sleepy-{second}"
        t = threading.Thread(target=sleepy, args=(second, thread_id))
        t.start()
        current_threads.append(t)
    
    for t in current_threads:
        t.join()

    tm2 = round(time.time() - start, 1)
    print('Time taken by sleep time fn: ', tm2, 'seconds\n')

    print(f"Total execution time of both fn: {tm1 + tm2} seconds")

if __name__ == "__main__":
    main(4, 6)


Thread fn-1 (fn) started.
Thread fn-2 (fn) started.
Thread fn-3 (fn) started.
Thread fn-4 (fn) started.
Thread fn-1 (fn) finished with sum: 333333283333335000000
Thread fn-2 (fn) finished with sum: 2666666466666670000000
Thread fn-3 (fn) finished with sum: 8999999550000005000000
Thread fn-4 (fn) finished with sum: 21333332533333340000000
Time taken by sum of square fn:  7.9 seconds

Thread sleepy-1 (sleepy) started, sleeping for 1 seconds.
Thread sleepy-2 (sleepy) started, sleeping for 2 seconds.
Thread sleepy-3 (sleepy) started, sleeping for 3 seconds.
Thread sleepy-4 (sleepy) started, sleeping for 4 seconds.
Thread sleepy-5 (sleepy) started, sleeping for 5 seconds.
Thread sleepy-1 (sleepy) finished.
Thread sleepy-2 (sleepy) finished.
Thread sleepy-3 (sleepy) finished.
Thread sleepy-4 (sleepy) finished.
Thread sleepy-5 (sleepy) finished.
Time taken by sleep time fn:  5.0 seconds

Total execution time of both fn: 12.9 seconds


#### explain

    When multiple threads, like fn-1 and fn-2, are executed on the same CPU core, they don't literally run at the same time in the strict sense. Instead, they share the CPU core's time through a process called time slicing or context switching.

    Time Slicing and Context Switching
    Time Slicing: The operating system's scheduler divides the CPU's time into small slices, allocating each thread a brief period to run. This time slice is usually in the order of milliseconds.
    Context Switching: When a thread's time slice ends, the operating system saves its current state (context) and switches to another thread. The new thread then starts or resumes execution. This switching happens very quickly, making it appear as though multiple threads are running simultaneously, even on a single core.
    Example Execution with fn-1 and fn-2
    Starting fn-1:

    The operating system gives fn-1 a time slice on the CPU core.
    fn-1 begins calculating the sum of squares for n=10,000,000.
    It might perform some iterations of the loop and accumulate part of the sum.
    Context Switch to fn-2:

    After fn-1 has used its time slice, the operating system may decide to switch to fn-2.
    The current state of fn-1 (including the value of its variables, current instruction, etc.) is saved.
    The CPU core is now allocated to fn-2, which begins calculating the sum of squares for n=20,000,000.
    fn-2 performs some iterations of its loop.
    Switching Back to fn-1:

    After fn-2 uses its time slice, the operating system might switch back to fn-1.
    The saved state of fn-1 is restored, and it resumes its calculation right where it left off.
    This cycle of switching continues, with both fn-1 and fn-2 getting turns to run on the CPU core.
    Parallel-like Execution:

    Because the switching happens so fast, usually many thousands of times per second, it creates the illusion that both threads are running simultaneously.
    In reality, they are taking turns, but this is done so efficiently that the performance gain is still significant.
    Visualization of Time Slicing
    Imagine a simplified timeline where each letter represents a small time slice:

    plaintext
    Copy code
    Time --->

    CPU Core: | fn-1 | fn-1 | fn-2 | fn-1 | fn-2 | fn-2 | fn-1 | fn-2 | ...

    fn-1:       Running             Running             Running         
    fn-2:                 Running             Running          Running  
    Key Points:
    Rapid Switching: The CPU rapidly switches between fn-1 and fn-2, allowing both to make progress almost simultaneously.
    Efficient Use of CPU: This method maximizes the use of the CPU core, keeping it busy and reducing idle time.
    Illusion of Parallelism: Even on a single core, this rapid switching gives the appearance of parallel execution, which is why threading can be effective even on systems with limited cores.
    Conclusion:
    Even though fn-1 and fn-2 are technically sharing a single CPU core, they appear to run concurrently because the operating system efficiently manages time slices through context switching. This allows both threads to progress in their calculations without one having to completely finish before the other can start.

### Threading when a function retruns some value

##### a sequantial approach

In [6]:
import time
import threading
import queue

#input:  [2, 3, 4,  5]
#output: [4, 9, 16, 25]

def fn(n):
    time.sleep(2)
    sq = n**n
    return sq

def main():
    start = time.time()

    nm = [2,3,4,5]
    out_lst = []
    for i in nm:
        out_lst.append(fn(i))
    
    print("Squares: ", out_lst)
    tm2 = round(time.time() - start,1)
    print('time taken by sum of square fn: ', tm2, 'seconds')
    return out_lst

if __name__ == "__main__":
    main()

    

Squares:  [4, 27, 256, 3125]
time taken by sum of square fn:  8.0 seconds


##### Threading approach

In [9]:
import time
import threading
import queue

#input:  [2, 3, 4,  5]
#output: [4, 9, 16, 25]

def fn(n, qu):
    time.sleep(2)
    sq = n**n
    qu.put(sq)        #return sq

def main():
    start = time.time()

    nm = [2,3,4,5]
    out_queue = queue.Queue()   #out_lst = []
    threads = []

    for number in nm:
        t = threading.Thread(target= fn, args=(number, out_queue))
        t.start()
        threads.append(t)
    
    for i in range(len(threads)):
        threads[i].join()
        
    
    
    result = []
    while not out_queue.empty():
        result.append(out_queue.get())


    print("Squares: ", result)
    tm2 = round(time.time() - start,1)
    print('time taken by sum of square fn: ', tm2, 'seconds')
    return out_queue

if __name__ == "__main__":
    main()

    

Squares:  [4, 256, 27, 3125]
time taken by sum of square fn:  2.0 seconds


##### queue concept refresher

In [17]:
from queue import Queue
q = Queue(maxsize = 3)
print(q.qsize()) 
q.put('a')
q.put('b')
q.put('c')
print(q.qsize(), q.full())
print(q.get())
print(q.qsize(), q.full())
q.put('e')
print(q.qsize(), q.full(), q.empty())


0
3 True
a
2 False
3 True False


#### alternative way to store and retrun the value

In [15]:
# using threadpooling method

from concurrent.futures import ThreadPoolExecutor
def fn(n):
    return n**n

def main():
    numbers = [2,3,4,5]
    Results = []

    with ThreadPoolExecutor(max_workers=4) as executor:
        futures = [executor.submit(fn, num) for num in numbers]

    for future in futures:
        Results.append(future.result())

    print("Square:",Results)



if __name__ == "__main__":
    main()

Square: [4, 27, 256, 3125]


In [22]:
# using dictonary

import threading

def fn(n, results):
    sq = n*n
    thread_id = threading.get_ident()
    print(thread_id)
    results[thread_id] = sq
    pass

def main():
    results = {}
    numbers = [4,5,6,7]

    threads = []

    for number in numbers:
        t = threading.Thread(target= fn , args= (number, results))
        t.start()
        threads.append(t)

    for thread in threads:
        thread.join()
    
    print("Squares: ", results.values())


if __name__== "__main__":
    main()

9780
14900
7772
14560
Squares:  dict_values([16, 25, 36, 49])


In [31]:
# using custom threading class

import threading

class fnClass(threading.Thread):
    def __init__(self, n):
        threading.Thread.__init__(self)
        self.n = n
        self.result = None

    def run(self):
        self.result = self.n * self.n
    



def main():
    threads = []
    numbers = [2, 3, 4, 5]

    for number in numbers:
        t = fnClass(number)
        t.start()
        threads.append(t)

    results = []
    for t in threads:
        t.join()
        results.append(t.result)

    print("Squares:", results)

if __name__ == "__main__":
    main()

    

Squares: [4, 9, 16, 25]


### different use cases of Threading

##### usecase1

In [ ]:
#IO operation handling

import threading
import time
import requests

# Custom thread class for downloading files
class DownloadThread(threading.Thread):
    def __init__(self, url, filename):
        threading.Thread.__init__(self)
        self.url = url
        self.filename = filename

    def run(self):
        response = requests.get(self.url)
        with open(self.filename, 'wb') as file:
            file.write(response.content)
        print(f"Downloaded {self.filename}")

def main():
    urls = [
        ("https://example.com/file1", "file1.txt"),
        ("https://example.com/file2", "file2.txt"),
        ("https://example.com/file3", "file3.txt"),
    ]
    threads = []

    for url, filename in urls:
        t = DownloadThread(url, filename)
        t.start()
        threads.append(t)

    for t in threads:
        t.join()

    print("All downloads completed.")

if __name__ == "__main__":
    main()


##### usecase2

In [37]:
#### prallal computation

import numpy as np

A = np.random.rand(3, 3)
B = np.random.rand(3, 3)
result = np.zeros((3, 3))
A, B, result

(array([[0.77066554, 0.63134172, 0.05582189],
        [0.85896099, 0.84747234, 0.67985402],
        [0.88105652, 0.30677149, 0.00311544]]),
 array([[0.14334451, 0.80528668, 0.22224675],
        [0.89451653, 0.62532771, 0.61644714],
        [0.61782207, 0.41474211, 0.63210282]]),
 array([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]]))

In [43]:
for i in range(A.shape[0]):
    break

i, A.shape[0]

(0, 3)

In [44]:
A[0]

array([0.77066554, 0.63134172, 0.05582189])

In [ ]:
np.dot()

In [ ]:
import threading
import numpy as np

# Custom thread class for matrix multiplication
class MatrixMultiplyThread(threading.Thread):
    def __init__(self, A, B, result, row):
        threading.Thread.__init__(self)
        self.A = A
        self.B = B
        self.result = result
        self.row = row

    def run(self):
        self.result[self.row] = np.dot(self.A[self.row], self.B)

def main():
    A = np.random.rand(3, 3)
    B = np.random.rand(3, 3)
    result = np.zeros((3, 3))

    threads = []
    for i in range(A.shape[0]):
        t = MatrixMultiplyThread(A, B, result, i)
        t.start()
        threads.append(t)

    for t in threads:
        t.join()

    print("Matrix A:\n", A)
    print("Matrix B:\n", B)
    print("Result of A * B:\n", result)

if __name__ == "__main__":
    main()


##### usecase3

In [ ]:
# handling background execution 

import threading
import time

# Custom thread class for background task
class BackgroundTask(threading.Thread):
    def __init__(self, interval):
        threading.Thread.__init__(self)
        self.interval = interval
        self.running = True

    def run(self):
        while self.running:
            print("Background task running...")
            time.sleep(self.interval)

    def stop(self):
        self.running = False

def main():
    background_task = BackgroundTask(interval=2)
    background_task.start()

    time.sleep(6)
    print("Main thread doing other tasks...")
    
    background_task.stop()
    background_task.join()
    print("Background task stopped.")

if __name__ == "__main__":
    main()


##### usecase4

In [ ]:
## handling concurrent user request

import threading
import socket

# Custom thread class to handle client connections
class ClientHandlerThread(threading.Thread):
    def __init__(self, client_socket, client_address):
        threading.Thread.__init__(self)
        self.client_socket = client_socket
        self.client_address = client_address

    def run(self):
        print(f"Connection from {self.client_address}")
        self.client_socket.sendall(b"Welcome to the server!\n")
        self.client_socket.close()

def main():
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind(('localhost', 9999))
    server_socket.listen(5)

    print("Server listening on port 9999...")
    while True:
        client_socket, client_address = server_socket.accept()
        handler = ClientHandlerThread(client_socket, client_address)
        handler.start()

if __name__ == "__main__":
    main()
